src/pycrunch/shoji.py

```diff
@@ -70,16 +70,44 @@ def __init__(self, session, catalog_url, **members):
         self.catalog_url = catalog_url
         catalog_url_absolute, frag = urllib.parse.urldefrag(catalog_url.absolute)
 
+        # Normalized keys stores a dictionary of shortest relative URLs
+        # obtained via URL.relative_to that maps to the arbitrary URL relative
+        # or absolute that the client submitted. We can use this to always
+        # key the tuples by relative URL regardless of what kind of URL the
+        # user asks to fetch an item by.
+        normalized_keys = {}
```

Comment:
would it be better for long-running sessions to have this as an LRU instead of a dict?